### DISCUSSION TITLE
formatEther formatting options (namely max decimals)

### DISCUSSION THREAD

**Discussion by SaulDope** [wevm/viem#955] [Idea / Feature Request]
When I'm doing math I want to keep the decimals of course, but when I want to display things to users I want to be able to truncate irrelevant decimals.

From an example on my site, I have a calculation that resulted in 39999999999999996 wei, or 0.039999999999999996 ether.

This is obviously ridiculous to show to users, realistically I would want to show something like 0.03999 or even 0.04.  I decided to go with the following simple truncate logic, you could likely do better if you offered more intelligent rounding options like RoundBank, RoundDown etc.

```
export const customFormatEther = (b: bigint, d: number) => {
  const es = formatEther(b)
  const est = es.split('.')
  if (est.length < 2) {
    return es
  }
  return est[0] + '.' + est[1].substring(0, d)
}
```

Should be a decent option for a new contributor to implement, I would be down to do so

**Comment by tmm** [comment]
For now, check out dnum's [`format`](https://github.com/bpierre/dnum#formatvalue-options) function.

**Comment by qantrepreneur** [comment]
For anyone looking, I wrote a solution:
```ts
export function customFormatEther(
  wei: bigint,
  maxDecimals: number = 18,
  rounding: "floor" | "ceil" | "round" = "round",
) {
  const power = 10n ** BigInt(18 - maxDecimals);
  const remainder = wei % power;
  const firstDecimal = remainder / 10n ** BigInt(18 - maxDecimals - 1);

  let b = wei - remainder;

  if (rounding === "ceil" || (rounding === "round" && firstDecimal >= 5)) {
    b += power;
  }

  return formatEther(b);
}
```

Usage:
```ts
customFormatEther(weiAmount, 4)
```

EDIT: simplified the method.
